sbin/vinum/vinum.8
.\"  Hey, Emacs, edit this file in -*- nroff-fill -*- mode
.\"-
.\" Copyright (c) 1997, 1998
.\"	Nan Yang Computer Services Limited.  All rights reserved.
.\"
.\"  This software is distributed under the so-called ``Berkeley
.\"  License'':
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the Company nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" This software is provided ``as is'', and any express or implied
.\" warranties, including, but not limited to, the implied warranties of
.\" merchantability and fitness for a particular purpose are disclaimed.
.\" In no event shall the company or contributors be liable for any
.\" direct, indirect, incidental, special, exemplary, or consequential
.\" damages (including, but not limited to, procurement of substitute
.\" goods or services; loss of use, data, or profits; or business
.\" interruption) however caused and on any theory of liability, whether
.\" in contract, strict liability, or tort (including negligence or
.\" otherwise) arising in any way out of the use of this software, even if
.\" advised of the possibility of such damage.
.\"
.\" $Id: vinum.8,v 1.48 2001/01/15 22:15:05 grog Exp $
.\" $FreeBSD: src/sbin/vinum/vinum.8,v 1.33.2.10 2002/12/29 16:35:38 schweikh Exp $
.\"
.Dd August 11, 2007
.Dt VINUM 8
.Os
.Sh NAME
.Nm vinum
.Nd Logical Volume Manager control program
.Sh SYNOPSIS
.Nm
.Op Ar command
.Op Fl options
.Sh COMMANDS
.Bl -tag -width indent
.It Ic attach Ar plex volume Op Cm rename
.It Xo
.Ic attach Ar subdisk plex
.Op Ar offset
.Op Cm rename
.Xc
Attach a plex to a volume, or a subdisk to a plex.
.It Xo
.Ic checkparity Ar plex
.Op Fl f
.Op Fl v
.Xc
Check the parity blocks of a RAID-4 or RAID-5 plex.
.It Xo
.Ic concat
.Op Fl f
.Op Fl n Ar name
.Op Fl v
.Ar drives
.Xc
Create a concatenated volume from the specified drives.
.It Xo
.Ic create
.Op Fl f
.Ar description-file
.Xc
Create a volume as described in
.Ar description-file .
.It Ic debug
Cause the volume manager to enter the kernel debugger.
.It Ic debug Ar flags
Set debugging flags.
.It Xo
.Ic detach
.Op Fl f
.Op Ar plex | subdisk
.Xc
Detach a plex or subdisk from the volume or plex to which it is attached.
.It Ic dumpconfig Op Ar drive ...
List the configuration information stored on the specified drives, or all drives
in the system if no drive names are specified.
.It Xo
.Ic info
.Op Fl v
.Op Fl V
.Xc
List information about volume manager state.
.It Xo
.Ic init
.Op Fl S Ar size
.Op Fl w
.Ar plex | subdisk
.Xc
.\" XXX
Initialize the contents of a subdisk or all the subdisks of a plex to all zeros.
.It Ic label Ar volume
Create a volume label.
.It Xo
.Ic l | list
.Op Fl r
.Op Fl s
.Op Fl v
.Op Fl V
.Op Ar volume | plex | subdisk
.Xc
List information about specified objects.
.It Xo
.Ic ld
.Op Fl r
.Op Fl s
.Op Fl v
.Op Fl V
.Op Ar volume
.Xc
List information about drives.
.It Xo
.Ic ls
.Op Fl r
.Op Fl s
.Op Fl v
.Op Fl V
.Op Ar subdisk
.Xc
List information about subdisks.
.It Xo
.Ic lp
.Op Fl r
.Op Fl s
.Op Fl v
.Op Fl V
.Op Ar plex
.Xc
List information about plexes.
.It Xo
.Ic lv
.Op Fl r
.Op Fl s
.Op Fl v
.Op Fl V
.Op Ar volume
.Xc
List information about volumes.
.It Ic makedev
Remake the device nodes in
.Pa /dev/vinum .
.It Xo
.Ic mirror
.Op Fl f
.Op Fl n Ar name
.Op Fl s
.Op Fl v
.Ar drives
.Xc
Create a mirrored volume from the specified drives.
.It Xo
.Ic move | mv
.Fl f
.Ar drive object ...
.Xc
Move the object(s) to the specified drive.
.It Ic printconfig Op Ar file
Write a copy of the current configuration to
.Ar file .
.It Ic quit
Exit the
.Nm
program when running in interactive mode.  Normally this would be done by
entering the
.Dv EOF
character.
.It Ic read Ar disk ...
Read the
.Nm
configuration from the specified disks.
.It Xo
.Ic rename Op Fl r
.Op Ar drive | subdisk | plex | volume
.Ar newname
.Xc
Change the name of the specified object.
.\" XXX
.\".It Ic replace Ar drive newdrive
.\"Move all the subdisks from the specified drive onto the new drive.
.It Xo
.Ic rebuildparity Ar plex Op Fl f
.Op Fl v
.Op Fl V
.Xc
Rebuild the parity blocks of a RAID-4 or RAID-5 plex.
.It Ic resetconfig
Reset the complete
.Nm
configuration.
.It Xo
.Ic resetstats
.Op Fl r
.Op Ar volume | plex | subdisk
.Xc
Reset statistics counters for the specified objects, or for all objects if none
are specified.
.It Xo
.Ic rm
.Op Fl f
.Op Fl r
.Ar volume | plex | subdisk
.Xc
Remove an object.
.It Ic saveconfig
Save
.Nm
configuration to disk after configuration failures.
.\" XXX
.\".It Xo
.\".Ic set
.\".Op Fl f
.\".Ar state
.\".Ar volume | plex | subdisk | disk
.\".Xc
.\"Set the state of the object to
.\".Ar state .
.It Ic setdaemon Op Ar value
Set daemon configuration.
.It Xo
.Ic setstate
.Ar state
.Op Ar volume | plex | subdisk | drive
.Xc
Set state without influencing other objects, for diagnostic purposes only.
.It Ic start
Read configuration from all vinum drives.
.It Xo
.Ic start
.Op Fl i Ar interval
.Op Fl S Ar size
.Op Fl w
.Ar volume | plex | subdisk
.Xc
Allow the system to access the objects.
.It Xo
.Ic stop
.Op Fl f
.Op Ar volume | plex | subdisk
.Xc
Terminate access to the objects, or stop
.Nm
if no parameters are specified.
.It Xo
.Ic stripe
.Op Fl f
.Op Fl n Ar name
.Op Fl v
.Ar drives
.Xc
Create a striped volume from the specified drives.
.El
.Sh DESCRIPTION
.Nm
is a utility program to communicate with the
.Xr vinum 4
logical volume
manager.
.Nm
is designed either for interactive use, when started without command line
arguments, or to execute a single command if the command is supplied on the
command line.  In interactive mode,
.Nm
maintains a command line history.
.Sh OPTIONS
.Nm
commands may optionally be followed by an option.  Any of the following options
may be specified with any command, but in some cases the options are ignored.
For example, the
.Ic stop
command ignores the
.Fl v
and
.Fl V
options.
.Bl -tag -width indent
.It Fl f
The
.Fl f
.Pq Dq force
option overrides safety checks.  Use with extreme care.  This option is for
emergency use only.  For example, the command
.Pp
.Dl rm -f myvolume
.Pp
removes
.Ar myvolume
even if it is open.  Any subsequent access to the volume will almost certainly
cause a panic.
.It Fl i Ar millisecs
When performing the
.Ic init
and
.Ic start
commands, wait
.Ar millisecs
milliseconds between copying each block.  This lowers the load on the system.
.It Fl n Ar name
Use the
.Fl n
option to specify a volume name to the simplified configuration commands
.Ic concat , mirror
and
.Ic stripe .
.It Fl r
The
.Fl r
.Pq Dq recursive
option is used by the list commands to display information not
only about the specified objects, but also about subordinate objects.  For
example, in conjunction with the
.Ic lv
command, the
.Fl r
option will also show information about the plexes and subdisks belonging to the
volume.
.It Fl s
The
.Fl s
.Pq Dq statistics
option is used by the list commands to display statistical information.  The
.Ic mirror
command also uses this option to specify that it should create striped plexes.
.It Fl S Ar size
The
.Fl S
option specifies the transfer size for the
.Ic init
and
.Ic start
commands.
.It Fl v
The
.Fl v
.Pq Dq verbose
option can be used to request more detailed information.
.It Fl V
The
.Fl V
.Pq Dq Very verbose
option can be used to request more detailed information than the
.Fl v
option provides.
.It Fl w
The
.Fl w
.Pq Dq wait
option tells
.Nm
to wait for completion of commands which normally run in the background, such as
.Ic init .
.El
.Sh COMMANDS IN DETAIL
.Nm
commands perform the following functions:
.Pp
.Bl -tag -width indent -compact
.It Ic attach Ar plex volume Op Cm rename
.It Xo
.Ic attach Ar subdisk plex
.Op Ar offset
.Op Cm rename
.Xc
.Nm Ic attach
inserts the specified plex or subdisk in a volume or plex.  In the case of a
subdisk, an offset in the plex may be specified.  If it is not, the subdisk will
be attached at the first possible location.  After attaching a plex to a
non-empty volume,
.Nm
reintegrates the plex.
.Pp
If the keyword
.Cm rename
is specified,
.Nm
renames the object (and in the case of a plex, any subordinate subdisks) to fit
in with the default
.Nm
naming convention.  To rename the object to any other name, use the
.Ic rename
command.
.Pp
A number of considerations apply to attaching subdisks:
.Bl -bullet
.It
Subdisks can normally only be attached to concatenated plexes.
.It
If a striped or RAID-5 plex is missing a subdisk (for example after drive
failure), it should be replaced by a subdisk of the same size only.
.It
In order to add further subdisks to a striped or RAID-5 plex, use the
.Fl f
(force) option.  This will corrupt the data in the plex.
.\"No other attachment of
.\"subdisks is currently allowed for striped and RAID-5 plexes.
.It
For concatenated plexes, the
.Ar offset
parameter specifies the offset in blocks from the beginning of the plex.  For
striped and RAID-5 plexes, it specifies the offset of the first block of the
subdisk: in other words, the offset is the numerical position of the subdisk
multiplied by the stripe size.  For example, in a plex with stripe size 271k,
the first subdisk will have offset 0, the second offset 271k, the third 542k,
etc.  This calculation ignores parity blocks in RAID-5 plexes.
.El
.Pp
.It Xo
.Ic checkparity
.Ar plex
.Op Fl f
.Op Fl v
.Xc
Check the parity blocks on the specified RAID-4 or RAID-5 plex.  This operation
maintains a pointer in the plex, so it can be stopped and later restarted from
the same position if desired.  In addition, this pointer is used by the
.Ic rebuildparity
command, so rebuilding the parity blocks need only start at the location where
the first parity problem has been detected.
.Pp
If the
.Fl f
flag is specified,
.Ic checkparity
starts checking at the beginning of the plex.  If the
.Fl v
flag is specified,
.Ic checkparity
prints a running progress report.
.Pp
.It Xo
.Ic concat
.Op Fl f
.Op Fl n Ar name
.Op Fl v
.Ar drives
.Xc
The
.Ic concat
command provides a simplified alternative to the
.Ic create
command for creating volumes with a single concatenated plex.  The largest
contiguous space available on each drive is used to create the subdisks for the
plexes.
.Pp
Normally, the
.Ic concat
command creates an arbitrary name for the volume and its components.  The name
is composed of the text
.Dq Li vinum
and a small integer, for example
.Dq Li vinum3 .
You can override this with the
.Fl n Ar name
option, which assigns the name specified to the volume.  The plexes and subdisks
are named after the volume in the default manner.
.Pp
There is no choice of name for the drives.  If the drives have already been
initialized as
.Nm
drives, the name remains.  Otherwise the drives are given names starting with
the text
.Dq Li vinumdrive
and a small integer, for example
.Dq Li vinumdrive7 .
As with the
.Ic create
command, the
.Fl f
option can be used to specify that a previous name should be overwritten.  The
.Fl v
is used to specify verbose output.
.Pp
See the section
.Sx SIMPLIFIED CONFIGURATION
below for some examples of this
command.
.Pp
.It Xo
.Ic create
.Op Fl f
.Ar description-file
.Xc
.Nm Ic create
is used to create any object.  In view of the relatively complicated
relationship and the potential dangers involved in creating a
.Nm
object, there is no interactive interface to this function.  If you do not
specify a file name,
.Nm
starts an editor on a temporary file.  If the environment variable
.Ev EDITOR
is set,
.Nm
starts this editor.  If not, it defaults to
.Nm vi .
See the section
.Sx CONFIGURATION FILE
below for more information on the format of
this file.
.Pp
Note that the
.Nm Ic create
function is additive: if you run it multiple times, you will create multiple
copies of all unnamed objects.
.Pp
Normally the
.Ic create
command will not change the names of existing
.Nm
drives, in order to avoid accidentally erasing them.  The correct way to dispose
of no longer wanted
.Nm
drives is to reset the configuration with the
.Ic resetconfig
command.  In some cases, however, it may be necessary to create new data on
.Nm
drives which can no longer be started.  In this case, use the
.Ic create Fl f
command.
.Pp
.It Ic debug
.Nm Ic debug ,
without any arguments, is used to enter the remote kernel debugger.  It is only
activated if
.Nm
is built with the
.Dv VINUMDEBUG
option.  This option will stop the execution of the operating system until the
kernel debugger is exited.  If remote debugging is set and there is no remote
connection for a kernel debugger, it will be necessary to reset the system and
reboot in order to leave the debugger.
.Pp
.It Ic debug Ar flags
Set a bit mask of internal debugging flags.  These will change without warning
as the product matures; to be certain, read the header file
.Pa /sys/dev/raid/vinum/vinumvar.h .
The bit mask is composed of the following values:
.Bl -tag -width indent
.It Dv DEBUG_ADDRESSES Pq No 1
Show buffer information during requests
.\".It Dv DEBUG_NUMOUTPUT Pq No 2
.\"Show the value of
.\".Va vp->v_numoutput .
.It Dv DEBUG_RESID Pq No 4
Go into debugger in
.Fn complete_rqe .
.It Dv DEBUG_LASTREQS Pq No 8
Keep a circular buffer of last requests.
.It Dv DEBUG_REVIVECONFLICT Pq No 16
Print info about revive conflicts.
.It Dv DEBUG_EOFINFO Pq No 32
Print information about internal state when returning an
.Dv EOF
on a striped plex.
.It Dv DEBUG_MEMFREE Pq No 64
Maintain a circular list of the last memory areas freed by the memory allocator.
.It Dv DEBUG_REMOTEGDB Pq No 256
Go into remote

sys/dev/raid/vinum/vinum.c
/*-
 * Copyright (c) 1997, 1998
 *	Nan Yang Computer Services Limited.  All rights reserved.
 *
 *  Written by Greg Lehey
 *
 *  This software is distributed under the so-called ``Berkeley
 *  License'':
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by Nan Yang Computer
 *      Services Limited.
 * 4. Neither the name of the Company nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * This software is provided ``as is'', and any express or implied
 * warranties, including, but not limited to, the implied warranties of
 * merchantability and fitness for a particular purpose are disclaimed.
 * In no event shall the company or contributors be liable for any
 * direct, indirect, incidental, special, exemplary, or consequential
 * damages (including, but not limited to, procurement of substitute
 * goods or services; loss of use, data, or profits; or business
 * interruption) however caused and on any theory of liability, whether
 * in contract, strict liability, or tort (including negligence or
 * otherwise) arising in any way out of the use of this software, even if
 * advised of the possibility of such damage.
 *
 * $Id: vinum.c,v 1.33 2001/01/09 06:19:15 grog Exp grog $
 * $FreeBSD: src/sys/dev/vinum/vinum.c,v 1.38.2.3 2003/01/07 12:14:16 joerg Exp $
 */

#define STATIC static					    /* nothing while we're testing XXX */

#include "vinumhdr.h"
#include <sys/sysmsg.h>					    /* for sync(2) */
#include <sys/poll.h>					    /* XXX: poll ops used in kq filters */
#include <sys/event.h>
#include <sys/udev.h>
#ifdef VINUMDEBUG
#include <sys/reboot.h>
int debug = 0;
extern int total_malloced;
extern int malloccount;
extern struct mc malloced[];
#endif
#include "request.h"

struct dev_ops vinum_ops =
{
	{ "vinum", 0, D_DISK },
	.d_open =	vinumopen,
	.d_close =	vinumclose,
	.d_read =	physread,
	.d_write =	physwrite,
	.d_ioctl =	vinumioctl,
	.d_kqfilter =	vinumkqfilter,
	.d_strategy =	vinumstrategy,
	.d_dump =	vinumdump,
	.d_psize =	vinumsize,
};

/* Called by main() during pseudo-device attachment. */
STATIC void vinumattach(void *);

STATIC int vinum_modevent(module_t mod, modeventtype_t type, void *unused);
STATIC void vinum_initconf(void);

struct _vinum_conf vinum_conf;				    /* configuration information */
cdev_t	vinum_super_dev;
cdev_t	vinum_wsuper_dev;
cdev_t	vinum_daemon_dev;

/*
 * Called by main() during pseudo-device attachment.  All we need
 * to do is allocate enough space for devices to be configured later, and
 * add devsw entries.
 */
static void
vinumattach(void *dummy)
{
    char *cp, *cp1, *cp2, **drives;
    int i, rv;
    struct volume *vol;

    /* modload should prevent multiple loads, so this is worth a panic */
    if ((vinum_conf.flags & VF_LOADED) != 0)
	panic("vinum: already loaded");

    log(LOG_INFO, "vinum: loaded\n");
    vinum_conf.flags |= VF_LOADED;			    /* we're loaded now */

    daemonq = NULL;					    /* initialize daemon's work queue */
    dqend = NULL;

#if 0
    dev_ops_add(&vinum_ops, 0, 0);
#endif

    vinum_initconf();

    /*
     * Create superdev, wrongsuperdev, and controld devices.
     */
    vinum_super_dev =	make_dev(&vinum_ops, VINUM_SUPERDEV,
				 UID_ROOT, GID_WHEEL, 0600,
				 VINUM_SUPERDEV_BASE);
    vinum_wsuper_dev =	make_dev(&vinum_ops, VINUM_WRONGSUPERDEV,
				 UID_ROOT, GID_WHEEL, 0600,
				 VINUM_WRONGSUPERDEV_BASE);
    vinum_daemon_dev = 	make_dev(&vinum_ops, VINUM_DAEMON_DEV,
				 UID_ROOT, GID_WHEEL, 0600,
				 VINUM_DAEMON_DEV_BASE);

    /*
     * See if the loader has passed us a disk to
     * read the initial configuration from.
     */
    if ((cp = kgetenv("vinum.drives")) != NULL) {
	for (cp1 = cp, i = 0, drives = NULL; *cp1 != '\0'; i++) {
	    cp2 = cp1;
	    while (*cp1 != '\0' && *cp1 != ',' && *cp1 != ' ')
		cp1++;
	    if (*cp1 != '\0')
		*cp1++ = '\0';
	    drives = krealloc(drives, (unsigned long)((i + 1) * sizeof(char *)),
			     M_TEMP, M_WAITOK);
	    drives[i] = cp2;
	}
	if (i == 0)
	    goto bailout;
	rv = vinum_scandisk(drives, i);
	if (rv)
	    log(LOG_NOTICE, "vinum_scandisk() returned %d", rv);
    bailout:
	kfree(drives, M_TEMP);
    }
    if ((cp = kgetenv("vinum.root")) != NULL) {
	for (i = 0; i < vinum_conf.volumes_used; i++) {
	    vol = &vinum_conf.volume[i];
	    if ((vol->state == volume_up)
		&& (strcmp (vol->name, cp) == 0) 
	    ) {
		rootdev = make_dev(&vinum_ops, i, UID_ROOT, GID_OPERATOR,
				0640, VINUM_BASE "vinumroot");
		udev_dict_set_cstr(rootdev, "subsystem", "raid");
		udev_dict_set_cstr(rootdev, "disk-type", "raid");
		log(LOG_INFO, "vinum: using volume %s for root device\n", cp);
		break;
	    }
	}
    }
}

/*
 * Check if we have anything open.  If confopen is != 0,
 * that goes for the super device as well, otherwise
 * only for volumes.
 *
 * Return 0 if not inactive, 1 if inactive.
 */
int
vinum_inactive(int confopen)
{
    int i;
    int can_do = 1;					    /* assume we can do it */

    if (confopen && (vinum_conf.flags & VF_OPEN))	    /* open by vinum(8)? */
	return 0;					    /* can't do it while we're open */
    lock_config();
    for (i = 0; i < vinum_conf.volumes_allocated; i++) {
	if ((VOL[i].state > volume_down)
	    && (VOL[i].flags & VF_OPEN)) {		    /* volume is open */
	    can_do = 0;
	    break;
	}
    }
    unlock_config();
    return can_do;
}

/*
 * Free all structures.
 * If cleardrive is 0, save the configuration; otherwise
 * remove the configuration from the drive.
 *
 * Before coming here, ensure that no volumes are open.
 */
void
free_vinum(int cleardrive)
{
    union daemoninfo di = { .nothing = 0 };
    int i;
    int drives_allocated = vinum_conf.drives_allocated;

    if (DRIVE != NULL) {
	if (cleardrive) {				    /* remove the vinum config */
	    for (i = 0; i < drives_allocated; i++)
		remove_drive(i);			    /* remove the drive */
	} else {					    /* keep the config */
	    for (i = 0; i < drives_allocated; i++)
		free_drive(&DRIVE[i]);			    /* close files and things */
	}
	Free(DRIVE);
    }
    while ((vinum_conf.flags & (VF_STOPPING | VF_DAEMONOPEN))
	== (VF_STOPPING | VF_DAEMONOPEN)) {		    /* at least one daemon open, we're stopping */
	queue_daemon_request(daemonrq_return, di);	    /* stop the daemon */
	tsleep(&vinumclose, 0, "vstop", 1);		    /* and wait for it */
    }
    if (SD != NULL) {
	for (i = 0; i < vinum_conf.subdisks_allocated; i++) {
	    struct sd *sd = &vinum_conf.sd[i];
	    if (sd->sd_dev) {
		destroy_dev(sd->sd_dev);
		sd->sd_dev = NULL;
	    }
	}
	Free(SD);
    }
    if (PLEX != NULL) {
	for (i = 0; i < vinum_conf.plexes_allocated; i++) {
	    struct plex *plex = &vinum_conf.plex[i];

	    if (plex->plex_dev) {
		destroy_dev(plex->plex_dev);
		plex->plex_dev = NULL;
	    }

	    if (plex->state != plex_unallocated) {	    /* we have real data there */
		if (plex->sdnos)
		    Free(plex->sdnos);
	    }
	}
	Free(PLEX);
    }
    if (VOL != NULL) {
	for (i = 0; i < vinum_conf.volumes_allocated; i++) {
	    struct volume *vol = &vinum_conf.volume[i];

	    if (vol->vol_dev) {
		destroy_dev(vol->vol_dev);
		vol->vol_dev = NULL;
	    }
	}
	Free(VOL);
    }
    bzero(&vinum_conf, sizeof(vinum_conf));
    vinum_initconf();
}

STATIC void
vinum_initconf(void)
{
    vinum_conf.physbufs = nswbuf_kva / 2 + 1;

    /* allocate space: drives... */
    DRIVE = (struct drive *) Malloc(sizeof(struct drive) * INITIAL_DRIVES);
    CHECKALLOC(DRIVE, "vinum: no memory\n");
    bzero(DRIVE, sizeof(struct drive) * INITIAL_DRIVES);
    vinum_conf.drives_allocated = INITIAL_DRIVES;
    vinum_conf.drives_used = 0;

    /* volumes, ... */
    VOL = (struct volume *) Malloc(sizeof(struct volume) * INITIAL_VOLUMES);
    CHECKALLOC(VOL, "vinum: no memory\n");
    bzero(VOL, sizeof(struct volume) * INITIAL_VOLUMES);
    vinum_conf.volumes_allocated = INITIAL_VOLUMES;
    vinum_conf.volumes_used = 0;

    /* plexes, ... */
    PLEX = (struct plex *) Malloc(sizeof(struct plex) * INITIAL_PLEXES);
    CHECKALLOC(PLEX, "vinum: no memory\n");
    bzero(PLEX, sizeof(struct plex) * INITIAL_PLEXES);
    vinum_conf.plexes_allocated = INITIAL_PLEXES;
    vinum_conf.plexes_used = 0;

    /* and subdisks */
    SD = (struct sd *) Malloc(sizeof(struct sd) * INITIAL_SUBDISKS);
    CHECKALLOC(SD, "vinum: no memory\n");
    bzero(SD, sizeof(struct sd) * INITIAL_SUBDISKS);
    vinum_conf.subdisks_allocated = INITIAL_SUBDISKS;
    vinum_conf.subdisks_used = 0;
}

STATIC int
vinum_modevent(module_t mod, modeventtype_t type, void *unused)
{
    switch (type) {
    case MOD_LOAD:
	vinumattach(NULL);
	return 0;					    /* OK */
    case MOD_UNLOAD:
	if (!vinum_inactive(1))				    /* is anything open? */
	    return EBUSY;				    /* yes, we can't do it */
	vinum_conf.flags |= VF_STOPPING;		    /* note that we want to stop */
	sys_sync(NULL, NULL);			    /* write out buffers */
	free_vinum(0);					    /* clean up */

	if (vinum_super_dev) {
	    destroy_dev(vinum_super_dev);
	    vinum_super_dev = NULL;
	}
	if (vinum_wsuper_dev) {
	    destroy_dev(vinum_wsuper_dev);
	    vinum_wsuper_dev = NULL;
	}
	if (vinum_daemon_dev) {
	    destroy_dev(vinum_daemon_dev);
	    vinum_daemon_dev = NULL;
	}

	sync_devs();
#ifdef VINUMDEBUG
	if (total_malloced) {
	    int i;
#ifdef INVARIANTS
	    int *poke;
#endif

	    for (i = 0; i < malloccount; i++) {
		if (debug & DEBUG_WARNINGS)		    /* want to hear about them */
		    log(LOG_WARNING,
			"vinum: exiting with %d bytes malloced from %s:%d\n",
			malloced[i].size,
			malloced[i].file,
			malloced[i].line);
#ifdef INVARIANTS
		poke = &((int *) malloced[i].address)
		    [malloced[i].size / (2 * sizeof(int))]; /* middle of the area */
		if (*poke == 0xdeadc0de)		    /* already freed */
		    log(LOG_ERR,
			"vinum: exiting with malloc table inconsistency at %p from %s:%d\n",
			malloced[i].address,
			malloced[i].file,
			malloced[i].line);
#endif
		Free(malloced[i].address);
	    }
	}
#endif
	dev_ops_remove_all(&vinum_ops);
	log(LOG_INFO, "vinum: unloaded\n");		    /* tell the world */
	return 0;
    default:
	break;
    }
    return 0;
}

moduledata_t vinum_mod =
{
    "vinum",
    (modeventhand_t) vinum_modevent,
    0
};
DECLARE_MODULE(vinum, vinum_mod, SI_SUB_RAID, SI_ORDER_MIDDLE);
MODULE_VERSION(vinum, 1);

/* ARGSUSED */
/* Open a vinum object */
int
vinumopen(struct dev_open_args *ap)
{
    cdev_t dev = ap->a_head.a_dev;
    int error;
    unsigned int index;
    struct volume *vol;
    struct plex *plex;
    struct sd *sd;
    int devminor;					    /* minor number */

    devminor = minor(dev);
    error = 0;
    /* First, decide what we're looking at */
    switch (DEVTYPE(dev)) {
    case VINUM_VOLUME_TYPE:
	index = Volno(dev);
	if (index >= vinum_conf.volumes_allocated)
	    return ENXIO;				    /* no such device */
	vol = &VOL[index];

	switch (vol->state) {
	case volume_unallocated:
	case volume_uninit:
	    return ENXIO;

	case volume_up:
	    vol->flags |= VF_OPEN;			    /* note we're open */
	    return 0;

	case volume_down:
	    return EIO;

	default:
	    return EINVAL;
	}

    case VINUM_PLEX_TYPE:
	if (Volno(dev) >= vinum_conf.volumes_allocated)
	    return ENXIO;
	/* FALLTHROUGH */

    case VINUM_RAWPLEX_TYPE:
	index = Plexno(dev);				    /* get plex index in vinum_conf */
	if (index >= vinum_conf.plexes_allocated)
	    return ENXIO;				    /* no such device */
	plex = &PLEX[index];

	switch (plex->state) {
	case plex_referenced:
	case plex_unallocated:
	    return EINVAL;

	default:
	    plex->flags |= VF_OPEN;			    /* note we're open */
	    return 0;
	}

    case VINUM_SD_TYPE:
	if ((Volno(dev) >= vinum_conf.volumes_allocated)    /* no such volume */
	||(Plexno(dev) >= vinum_conf.plexes_allocated))	    /* or no such plex */
	    return ENXIO;				    /* no such device */

	/* FALLTHROUGH */

    case VINUM_RAWSD_TYPE:
	index = Sdno(dev);				    /* get the subdisk number */
	if ((index >= vinum_conf.subdisks_allocated)	    /* not a valid SD entry */
	||(SD[index].state < sd_init))			    /* or SD is not real */
	    return ENXIO;				    /* no such device */
	sd = &SD[index];

	/*
	 * Opening a subdisk is always a special operation, so we
	 * ignore the state as long as it represents a real subdisk
	 */
	switch (sd->state) {
	case sd_unallocated:
	case sd_uninit:
	    return EINVAL;

	default:
	    sd->flags |= VF_OPEN;			    /* note we're open */
	    return 0;
	}

    case VINUM_SUPERDEV_TYPE:
	error = priv_check_cred(ap->a_cred, PRIV_ROOT, 0);  /* are we root? */
	if (error == 0) {				    /* yes, can do */
	    if (devminor == VINUM_DAEMON_DEV)		    /* daemon device */
		vinum_conf.flags |= VF_DAEMONOPEN;	    /* we're open */
	    else if (devminor == VINUM_SUPERDEV)
		vinum_conf.flags |= VF_OPEN;		    /* we're open */
	    else
		error = ENODEV;				    /* nothing, maybe a debug mismatch */
	}
	return error;

	/* Vinum drives are disks.  We already have a disk
	 * driver, so don't handle them here */
    case VINUM_DRIVE_TYPE:
    default:
	return ENODEV;					    /* don't know what to do with these */
    }
}

/* ARGSUSED */
int
vinumclose(struct dev_close_args *ap)
{
    cdev_t dev = ap->a_head.a_dev;
    unsigned int index;
    struct volume *vol;
    int devminor;

    devminor = minor(dev);
    index = Volno(dev);
    /* First, decide what we're looking at */
    switch (DEVTYPE(dev)) {
    case VINUM_VOLUME_TYPE:
	if (index >= vinum_conf.volumes_allocated)
	    return ENXIO;				    /* no such device */
	vol = &VOL[index];

	switch (vol->state) {
	case volume_unallocated:
	case volume_uninit:
	    return ENXIO;

	case volume_up:
	    vol->flags &= ~VF_OPEN;			    /* reset our flags */
	    return 0;

	case volume_down:
	    return EIO;

	default:
	    return EINVAL;
	}

    case VINUM_PLEX_TYPE:
	if (Volno(dev) >= vinum_conf.volumes_allocated)
	    return ENXIO;
	/* FALLTHROUGH */

    case VINUM_RAWPLEX_TYPE:
	index = Plexno(dev);				    /* get plex index in vinum_conf */
	if (index >= vinum_conf.plexes_allocated)
	    return ENXIO;				    /* no such device */
	PLEX[index].flags &= ~VF_OPEN;			    /* reset our flags */
	return 0;

    case VINUM_SD_TYPE:
	if ((Volno(dev) >= vinum_conf.volumes_allocated) || /* no such volume */
	    (Plexno(dev) >= vinum_conf.plexes_allocated))   /* or no such plex */
	    return ENXIO;				    /* no such device */
	/* FALLTHROUGH */

    case VINUM_RAWSD_TYPE:
	index = Sdno(dev);				    /* get the subdisk number */
	if (index >= vinum_conf.subdisks_allocated)
	    return ENXIO;				    /* no such device */
	SD[index].flags &= ~VF_OPEN;			    /* reset our flags */
	return 0;

    case VINUM_SUPERDEV_TYPE:
	/*
	 * don't worry about whether we're root:
	 * nobody else would get this far.
	 */
	if (devminor == VINUM_SUPERDEV)			    /* normal superdev */
	    vinum_conf.flags &= ~VF_OPEN;		    /* no longer open */
	else if (devminor == VINUM_DAEMON_DEV) {	    /* the daemon device */
	    vinum_conf.flags &= ~VF_DAEMONOPEN;		    /* no longer open */
	    if (vinum_conf.flags & VF_STOPPING)		    /* we're stopping, */
		wakeup(&vinumclose);			    /* we can continue stopping now */
	}
	return 0;

    case VINUM_DRIVE_TYPE:
    default:
	return ENODEV;					    /* don't know what to do with these */
    }
}

/* size routine */
int
vinumsize(struct dev_psize_args *ap)
{
    cdev_t dev = ap->a_head.a_dev;
    struct volume *vol;

    vol = &VOL[Volno(dev)];

    if (vol->state == volume_up) {
	ap->a_result = (int64_t)vol->size;
	return(0);
    } else {
	return(ENXIO);
    }
}

int
vinumdump(struct dev_dump_args *ap)
{
    /* Not implemented. */
    return ENXIO;
}

void
vinumfilt_detach(struct knote *kn) {}

int
vinumfilt_rd(struct knote *kn, long hint)
{
    cdev_t dev = (cdev_t)kn->kn_hook;

    if (seltrue(dev, POLLIN | POLLRDNORM))
        return (1);

    return (0);
}

int
vinumfilt_wr(struct knote *kn, long hint)
{
    /* Writing is always OK */
    return (1);
}

struct filterops vinumfiltops_rd =
    { FILTEROP_ISFD, NULL, vinumfilt_detach, vinumfilt_rd };
struct filterops vinumfiltops_wr =
    { FILTEROP_ISFD, NULL, vinumfilt_detach, vinumfilt_wr };

int
vinumkqfilter(struct dev_kqfilter_args *ap)
{
    if (ap->a_kn->kn_filter == EVFILT_READ) {
        ap->a_kn->kn_fop = &vinumfiltops_rd;
        ap->a_kn->kn_hook = (caddr_t)ap->a_head.a_dev;
        ap->a_result = 0;
    } else if (ap->a_kn->kn_filter == EVFILT_WRITE) {
	ap->a_kn->kn_fop = &vinumfiltops_wr;
        ap->a_result = 0;
    } else {
        ap->a_result = EOPNOTSUPP;
    }

    return (0);
}
